stonesoup/models/measurement/nonlinear.py

```diff
@@ -517,7 +552,322 @@ def function(self, state, noise=False, **kwargs):
+
+    def __init__(self, *args, **kwargs):
+        """
+        Ensure that the translation offset is initiated
+        """
+        super().__init__(*args, **kwargs)
+        # Set values to defaults if not provided
+        if self.translation_offset is None:
+            self.translation_offset = StateVector([0] * 3)
+
+        if self.velocity is None:
+            self.velocity = StateVector([0] * 3)
+
+    @property
+    def ndim_meas(self) -> int:
+        """ndim_meas getter method
+
+        Returns
+        -------
+        :class:`int`
+            The number of measurement dimensions
+        """
+
+        return 4
+
+    def function(self, state, noise=False, **kwargs) -> StateVector:
+        r"""Model function :math:`h(\vec{x}_t,\vec{v}_t)`
+
+        Parameters
+        ----------
+        state: :class:`~.StateVector`
+            An input state vector for the target
+
+        noise: :class:`numpy.ndarray` or bool
+            An externally generated random process noise sample (the default is
+            `False`, in which case no noise will be added
+            if 'True', the output of :meth:`~.Model.rvs` is added)
+
+        Returns
+        -------
+        :class:`numpy.ndarray` of shape (:py:attr:`~ndim_state`, 1)
+            The model function evaluated given the provided time interval.
+        """
+
+        if isinstance(noise, bool) or noise is None:
+            if noise:
+                noise = self.rvs()
+            else:
+                noise = 0
+
+        # Account for origin offset in position to enable range and angles to be determined
+        xyz_pos = state.state_vector[self.mapping] - self.translation_offset
+
+        # Rotate coordinates based upon the sensor_velocity
+        xyz_rot = self._rotation_matrix @ xyz_pos
+
+        # Convert to Spherical
+        rho, phi, theta = cart2sphere(*xyz_rot[:, 0])
+
+        # Determine the net velocity component in the engagement
+        xyz_vel = state.state_vector[self.velocity_mapping, :] - self.velocity
+
+        # Use polar to calculate range rate
+        rr = np.dot(xyz_pos[:, 0], xyz_vel[:, 0]) / np.linalg.norm(xyz_pos)
+
+        return StateVector([[Elevation(phi)],
+                            [Bearing(theta)],
+                            [rho],
+                            [rr]]) + noise
+
+    def inverse_function(self, detection, **kwargs) -> StateVector:
+        phi, theta, rho, rho_rate = detection.state_vector
+
+        [x, y, z] = sphere2cart(rho, theta, phi)
+        # because only rho_rate is known, only the components in
+        # x,y and z of the range rate can be found.
+        x_rate = np.cos(phi) * np.cos(theta) * rho_rate
+        y_rate = np.cos(phi) * np.sin(theta) * rho_rate
+        z_rate = np.sin(phi) * rho_rate
+
+        inv_rotation_matrix = inv(self._rotation_matrix)
+
+        out_vector = StateVector([[0.], [0.], [0.], [0.], [0.], [0.]])
+        out_vector[self.mapping] = x, y, z
```

Comment:
Some of these mappings may break, with tuple mappings: `out_vector[self.mapping, :]` would be better,

Reply:
modified